programs/eden-tester/main.cpp

```diff
@@ -0,0 +1,1141 @@
+      if (state.backend.get_module().tables.size() < 0 ||
+          state.backend.get_module().tables[0].table.size() < cb_alloc)
+         throw std::runtime_error("cb_alloc is out of range");
+      auto result = state.backend.get_context().execute(  //
+          this, eosio::vm::jit_visitor(42), state.backend.get_module().tables[0].table[cb_alloc],
+          cb_alloc_data, size);
+      if (!result || !result->is_a<eosio::vm::i32_const_t>())
+         throw std::runtime_error("cb_alloc returned incorrect type");
+      char* begin = state.wa.get_base_ptr<char>() + result->to_ui32();
+      check_bounds(begin, size);
+      return begin;
+   }
+
+   template <typename T>
+   void set_data(uint32_t cb_alloc_data, uint32_t cb_alloc, const T& data)
+   {
+      memcpy(alloc(cb_alloc_data, cb_alloc, data.size()), data.data(), data.size());
+   }
+
+   void tester_abort() { throw std::runtime_error("called tester_abort"); }
+
+   void eosio_exit(int32_t) { throw std::runtime_error("called eosio_exit"); }
+
+   void eosio_assert_message(bool condition, span<const char> msg)
+   {
+      if (!condition)
+         throw ::assert_exception(span_str(msg));
+   }
+
+   void prints_l(span<const char> str) { std::cerr.write(str.data(), str.size()); }
+
+   void tester_get_arg_counts(wasm_ptr<uint32_t> argc, wasm_ptr<uint32_t> argv_buf_size)
+   {
+      uint32_t size = 0;
+      for (auto& a : state.args)
+         size += a.size() + 1;
+      *argc = state.args.size();
+      *argv_buf_size = size;
+   };
+
+   // uint8_t** argv, uint8_t* argv_buf
+   void tester_get_args(uint32_t argv, uint32_t argv_buf)
+   {
+      auto* memory = state.backend.get_context().linear_memory();
+      auto get_argv = [&]() -> uint32_t& { return *reinterpret_cast<uint32_t*>(memory + argv); };
+      auto get_argv_buf = [&]() -> uint8_t& {
+         return *reinterpret_cast<uint8_t*>(memory + argv_buf);
+      };
+
+      for (auto& a : state.args)
+      {
+         get_argv() = argv_buf;
+         argv += 4;
+         for (auto ch : a)
+         {
+            get_argv_buf() = ch;
+            ++argv_buf;
+         }
+         get_argv_buf() = 0;
+         ++argv_buf;
+      }
+   };
+
+   int32_t clock_gettime(int32_t id, void* data)
+   {
+      check_bounds((char*)data, 8);
+      std::chrono::nanoseconds result;
+      if (id == 0)
+      {  // CLOCK_REALTIME
+         result = std::chrono::system_clock::now().time_since_epoch();
+      }
+      else if (id == 1)
+      {  // CLOCK_MONOTONIC
+         result = std::chrono::steady_clock::now().time_since_epoch();
+      }
+      else
+      {
+         return -1;
+      }
+      int32_t sec = result.count() / 1000000000;
+      int32_t nsec = result.count() % 1000000000;
+      fc::datastream<char*> ds((char*)data, 8);
+      fc::raw::pack(ds, sec);
+      fc::raw::pack(ds, nsec);
+      return 0;
+   }
+
+   int32_t open_file(span<const char> filename, span<const char> mode)
```

Comment:
The file handling here should be integrated with WASI.

Reply:
done